src/clingexplaid/mus/core_computer.py

```diff
@@ -70,32 +75,52 @@ def _compute_single_minimal(self, assumptions: Optional[AssumptionSet] = None) -
         if satisfiable:
             return UnsatisfiableSubset(set())
 
-        mus_members: Set[Assumption] = set()
         working_set = set(assumptions)
 
         for assumption in assumptions:
             # remove the current assumption from the working set
             working_set.remove(assumption)
 
-            satisfiable, _, _ = self._solve(assumptions=working_set.union(mus_members))
+            satisfiable, _, _ = self._solve(assumptions=working_set.union(self._shrinking_progress))
             # if the working set now becomes satisfiable without the assumption it is added to the mus_members
             if satisfiable:
-                mus_members.add(assumption)
+                self._shrinking_progress.add(assumption)
                 # every time we discover a new mus member we also check if all currently found mus members already
                 # suffice to make the instance unsatisfiable. If so we can stop the search sice we found our mus.
-                if not self._solve(assumptions=mus_members)[0]:
+                if not self._solve(assumptions=self._shrinking_progress)[0]:
                     break
 
-        return UnsatisfiableSubset(mus_members, minimal=True)
-
-    def shrink(self, assumptions: Optional[AssumptionSet] = None) -> UnsatisfiableSubset:
+            # Do a very short wait every time to allow for async timeout interruption
+            await asyncio.sleep(0.0000001)
+
+        return UnsatisfiableSubset(self._shrinking_progress, minimal=True)
+
+    async def _shrink_unsatisfiable_subset(
+        self, assumptions: Optional[AssumptionSet] = None, timeout: Optional[float] = None
+    ) -> UnsatisfiableSubset:
+        timeout_occurred = False
+        try:
+            await asyncio.wait_for(self._compute_single_minimal(assumptions=assumptions), timeout=timeout)
+        except TimeoutError:
+            timeout_occurred = True
+            warnings.warn(
+                "Timeout encountered while computing unsatisfiable subset, "
+                "intermediate unsatisfiable subset is returned"
+            )
+        # TODO: Theoretically even if a timeout occurred it could be the MUS but it's not guaranteed
```

Comment:
I think this is fine, it is like in clingo, you might be giving the optimal, but if the optimality is not proven you still get false.